Migrate to AI SDK v5 with Zod v4 Compatibility

bolechen opened the pull request:
# Migrate to AI SDK v5 with Zod v4 Compatibility

## Summary
This PR migrates the provider from AI SDK to v5, implementing the v2 specification interface and adding Zod v4 compatibility.

## Changes
✅ Migrate all models (chat, completion, embedding) to AI SDK v5 v2 interface
✅ Add Zod v4 compatibility while maintaining v3 support (^3.25.0 || ^4.0.0)
✅ Refactor tool response handling in convertToQwenChatMessages
✅ Update all tests to v2 specification using MSW for mocking
✅ Add prepack script to support git-based dependencies
✅ Code style improvements (consistent semicolons)

## Testing
✅ All tests passing (pnpm test)
✅ Lint checks passing (pnpm lint)
✅ Type checks passing (pnpm type-check)
✅ Validated in production project

## Notes

- Most code was AI-assisted but has been thoroughly tested locally
- Breaking changes are minimal due to backward-compatible peer dependencies
- Ready for production use


Close #8

bolechen committed:
feat: add MSW dependency and update test files for V2 API migration

- Added `msw` as a dependency for mocking API requests.
- Updated test files to utilize `createTestServer` for V2 API compatibility.
- Adjusted imports and response handling in tests for `QwenChatLanguageModel`, `QwenCompletionLanguageModel`, and `QwenEmbeddingModel`.

bolechen committed:
test: migrate all tests to AI SDK v5 specification v2

Completed comprehensive migration of all test suites to v2 API format,
maintaining 100% test coverage with zero functionality loss.

Changes:
- Migrated chat model tests (44 tests) to v2 format
- Migrated completion model tests (22 tests) to v2 format
- Migrated embedding model tests (6 tests) to v2 format
- Implemented custom fetch mocking pattern to avoid DataCloneError
- Updated all assertions to match v2 API response format

V2 API adaptations:
- Changed from text/toolCalls to content[] array format
- Updated usage fields: inputTokens/outputTokens (was promptTokens/completionTokens)
- Removed mode parameter; tools/toolChoice now direct parameters
- Changed rawResponse to response
- Updated stream parts: text-delta now uses 'delta' field (was 'textDelta')
- Added text-start/text-end, reasoning-start/reasoning-end patterns
- Changed providerMetadata to providerOptions

bolechen committed:
refactor: migrate Qwen models and tests to version 2 interface

- Updated all relevant files to utilize the new LanguageModelV2 interface, including changes to method signatures and response formats.
- Refactored test cases to align with the v2 API, ensuring compatibility and maintaining test coverage.
- Replaced `providerMetadata` with `providerOptions` across the codebase.
- Adjusted handling of content types, specifically changing `image` to `file` and updating associated properties.
- Removed deprecated `prepareTools` utility as part of the migration process.

bolechen committed:
fix: refine tool response handling in convertToQwenChatMessages

- Updated the content extraction logic for tool responses to accommodate both text and JSON output formats.
- Ensured that the content is correctly formatted based on the output type, improving the accuracy of message conversion.

bolechen committed:
style(project): add semicolons to all TypeScript files

This change adds missing semicolons to all TypeScript files
in the project to maintain consistent code style. The addition
of semicolons improves code readability and aligns with
standard TypeScript conventions.

style(project): add semicolons to all TypeScript files

This change adds missing semicolons to all TypeScript files
in the project to maintain consistent code style. The addition
of semicolons improves code readability and aligns with
standard TypeScript conventions.

bolechen committed:
Add Zod v4 compatibility and update peer dependencies

Updated peerDependencies and devDependencies to support both Zod v3.25.0+ and v4.0.0. Updated documentation to clarify Zod compatibility. Refactored type usage in qwen-error.ts to use z.ZodType for compatibility with both Zod versions. Added detailed JSDoc for getArgs options in qwen-chat-language-model.ts.

bolechen committed:
chore(build): add prepack script to ensure built artifacts

Add prepack script that runs the build command before packaging. This
ensures that distribution files are always up to date when the package
is packed for publication, preventing outdated builds from being
distributed.

bolechen committed:
refactor(tests): reorganize test files into __tests__ directory

Move all test files from src root to dedicated __tests__ directory to
improve project structure and follow common testing conventions. This
separation makes the codebase more organized and aligns with standard
JavaScript/TypeScript project layouts for better maintainability.

bolechen commented:
`pnpm test` && `pnpm lint` && `pnpm coverage`

All passed, plz have a reiview @Younis-Ahmed

bolechen commented:
You can test this branch in your project before it's merged by updating your package.json:

```
  "dependencies": {
    "qwen-ai-provider": "github:bolechen/qwen-ai-provider#feat/ai-sdk-v5"
  }
}
```

bolechen committed:
feat: enhance tool handling and error management in QwenChatLanguageModel

- Added support for logging warnings when unsupported tools are encountered.
- Implemented tests to ensure proper error handling for unknown content types and unsupported roles in assistant messages.
- Enhanced conversion logic for tool role outputs to avoid unnecessary JSON stringification, improving output consistency.

bolechen committed:
fix: improve error handling in QwenChat and QwenCompletion models

- Updated error handling logic to correctly process provider error objects streamed as chunks.
- Enhanced tests to verify proper handling of error responses and unsupported tools in the QwenChatLanguageModel.
- Added new test cases to ensure robustness against invalid response data in the QwenEmbeddingModel.
:

bolechen committed:
fix(completion): convert system message content parts to text

- Map system content parts to text to avoid [object Object] in prompt
- Throw on non-text system parts to surface unsupported usage
- Add test for system parts array in completion prompt

bolechen committed:
fix(peer-deps): tighten zod peer range to match provider-utils

- Set peer zod to ^3.25.76 || ^4.1.8 (intersection)
- Bump dev zod to ^4.1.8 to align
- Update README Zod Compatibility notes

thejozika commented:
@Younis-Ahmed is this repo still maintained?

SirBernardPhilip commented:
> @Younis-Ahmed is this repo still maintained?

It seems abandoned, if I were you I'd push your fork to the npm registry under a new name or something

bolechen committed:
refactor: migrate Qwen models to version 2 interface and update dependencies (#2)

* refactor: migrate Qwen models to version 2 interface and update dependencies

- Updated QwenChatLanguageModel, QwenCompletionLanguageModel, and QwenEmbeddingModel to implement LanguageModelV2.
- Adjusted method signatures and response formats to align with the new version.
- Updated dependencies for @ai-sdk/provider and @ai-sdk/provider-utils to their latest versions.

* feat: add MSW dependency and update test files for V2 API migration

- Added `msw` as a dependency for mocking API requests.
- Updated test files to utilize `createTestServer` for V2 API compatibility.
- Adjusted imports and response handling in tests for `QwenChatLanguageModel`, `QwenCompletionLanguageModel`, and `QwenEmbeddingModel`.

* test: migrate all tests to AI SDK v5 specification v2

Completed comprehensive migration of all test suites to v2 API format,
maintaining 100% test coverage with zero functionality loss.

Changes:
- Migrated chat model tests (44 tests) to v2 format
- Migrated completion model tests (22 tests) to v2 format
- Migrated embedding model tests (6 tests) to v2 format
- Implemented custom fetch mocking pattern to avoid DataCloneError
- Updated all assertions to match v2 API response format

V2 API adaptations:
- Changed from text/toolCalls to content[] array format
- Updated usage fields: inputTokens/outputTokens (was promptTokens/completionTokens)
- Removed mode parameter; tools/toolChoice now direct parameters
- Changed rawResponse to response
- Updated stream parts: text-delta now uses 'delta' field (was 'textDelta')
- Added text-start/text-end, reasoning-start/reasoning-end patterns
- Changed providerMetadata to providerOptions

* refactor: migrate Qwen models and tests to version 2 interface

- Updated all relevant files to utilize the new LanguageModelV2 interface, including changes to method signatures and response formats.
- Refactored test cases to align with the v2 API, ensuring compatibility and maintaining test coverage.
- Replaced `providerMetadata` with `providerOptions` across the codebase.
- Adjusted handling of content types, specifically changing `image` to `file` and updating associated properties.
- Removed deprecated `prepareTools` utility as part of the migration process.

* fix: refine tool response handling in convertToQwenChatMessages

- Updated the content extraction logic for tool responses to accommodate both text and JSON output formats.
- Ensured that the content is correctly formatted based on the output type, improving the accuracy of message conversion.

* style(project): add semicolons to all TypeScript files

This change adds missing semicolons to all TypeScript files
in the project to maintain consistent code style. The addition
of semicolons improves code readability and aligns with
standard TypeScript conventions.

style(project): add semicolons to all TypeScript files

This change adds missing semicolons to all TypeScript files
in the project to maintain consistent code style. The addition
of semicolons improves code readability and aligns with
standard TypeScript conventions.

* Add Zod v4 compatibility and update peer dependencies

Updated peerDependencies and devDependencies to support both Zod v3.25.0+ and v4.0.0. Updated documentation to clarify Zod compatibility. Refactored type usage in qwen-error.ts to use z.ZodType for compatibility with both Zod versions. Added detailed JSDoc for getArgs options in qwen-chat-language-model.ts.

* chore(build): add prepack script to ensure built artifacts

Add prepack script that runs the build command before packaging. This
ensures that distribution files are always up to date when the package
is packed for publication, preventing outdated builds from being
distributed.

* refactor(tests): reorganize test files into __tests__ directory

Move all test files from src root to dedicated __tests__ directory to
improve project structure and follow common testing conventions. This
separation makes the codebase more organized and aligns with standard
JavaScript/TypeScript project layouts for better maintainability.

* feat: enhance tool handling and error management in QwenChatLanguageModel

- Added support for logging warnings when unsupported tools are encountered.
- Implemented tests to ensure proper error handling for unknown content types and unsupported roles in assistant messages.
- Enhanced conversion logic for tool role outputs to avoid unnecessary JSON stringification, improving output consistency.

* fix: improve error handling in QwenChat and QwenCompletion models

- Updated error handling logic to correctly process provider error objects streamed as chunks.
- Enhanced tests to verify proper handling of error responses and unsupported tools in the QwenChatLanguageModel.
- Added new test cases to ensure robustness against invalid response data in the QwenEmbeddingModel.
:

* fix(completion): convert system message content parts to text

- Map system content parts to text to avoid [object Object] in prompt
- Throw on non-text system parts to surface unsupported usage
- Add test for system parts array in completion prompt

* fix(peer-deps): tighten zod peer range to match provider-utils

- Set peer zod to ^3.25.76 || ^4.1.8 (intersection)
- Bump dev zod to ^4.1.8 to align
- Update README Zod Compatibility notes

* chore(package): update package metadata and structure

bolechen commented:
> > @Younis-Ahmed is this repo still maintained?
> 
> It seems abandoned, if I were you I'd push your fork to the npm registry under a new name or something

https://www.npmjs.com/package/qwen-ai-provider-v5

```
pnpm add qwen-ai-provider-v5
```
have to try.

CatchMeX commented:
好像有点问题 这个pr对应的npm包和ai-sdk-v5一起使用时 会报错:toolCall.input.trim is not a function @bolechen

bolechen commented:
> 好像有点问题 这个pr对应的npm包和ai-sdk-v5一起使用时 会报错:toolCall.input.trim is not a function @bolechen


我自己的项目线上用了半个多月
暂时没有发现问题啊
用的是 v5 那个包吗
能不能做个最小化复现环境我看看
谢谢反馈

bolechen committed:
fix(qwen): ensure tool call input is stringified for AI SDK v5 compatibility (#3)

* fix(qwen): ensure tool call input is stringified for AI SDK v5 compatibility

Fixed QwenChatLanguageModel to emit tool-call stream parts with string
input payloads to match the AI SDK v5 LanguageModelV2ToolCall contract.
This resolves toolCall.input.trim is not a function errors when using
qwen-ai-provider-v5 together with ai-sdk v5 tool execution. Updated
tests to assert string input for tool calls in both streaming and
non-streaming modes.

* chore(package): update version to 1.0.2

This commit updates the package version from 1.0.1 to 1.0.2 to reflect
the latest changes and prepare for the next release cycle.

bolechen commented:
@CatchMeX 麻烦试下 1.0.2 这个版本

CatchMeX commented:
> @CatchMeX 麻烦试下 1.0.2 这个版本

ok 可以了 感谢

bolechen committed:
chore: Clean up repository configuration and documentation

wip

bolechen committed:
chore(deps): update dependencies and package configuration

Update package dependencies to latest versions including AI SDK v2 and
Zod v4. Add commonjs module type and update description to reflect Vercel
AI Provider v5 compatibility. Reorganize publishConfig section for better
readability while maintaining same publishing settings.
